File: realtime-macos/Sources/PAPAluLive/WindowScale.swift
struct WindowScale {
    private static let minimum = 0.5
    private static let maximum = 2.0
    private static let defaultFactor = 1.0
    private static let step = 0.1

    private(set) var factor = Self.defaultFactor

    mutating func increase() {
        factor = min(Self.maximum, rounded(factor + Self.step))
    }

    mutating func decrease() {
        factor = max(Self.minimum, rounded(factor - Self.step))
    }

    mutating func reset() {
        factor = Self.defaultFactor
    }

    private func rounded(_ value: Double) -> Double {
        (value * 10).rounded() / 10
    }
}

File: realtime-windows/PAPAluLive.Core/WindowScale.cs
namespace PAPAluLive.Core;

public sealed class WindowScale
{
    public const double Minimum = 0.5;
    public const double Maximum = 2.0;
    public const double DefaultFactor = 1.0;
    public const double Step = 0.1;

    public double Factor { get; private set; }

    public WindowScale(double factor = DefaultFactor)
    {
        Factor = Clamp(factor);
    }

    public void Increase() => Factor = Math.Min(Maximum, Round(Factor + Step));

    public void Decrease() => Factor = Math.Max(Minimum, Round(Factor - Step));

    public void Reset() => Factor = DefaultFactor;

    public void SetFactor(double value) => Factor = Clamp(value);

    private static double Round(double value) => Math.Round(value * 10) / 10;

    private static double Clamp(double value) => double.IsFinite(value)
        ? Math.Clamp(value, Minimum, Maximum)
        : DefaultFactor;
}
